drivers/staging/vc04_services/bcm2835-camera/bcm2835-camera.c

```diff
@@ -554,7 +554,8 @@ static int start_streaming(struct vb2_queue *vq, unsigned int count)
 			vchiq_mmal_port_return_buffers(dev->instance,
 						       dev->capture.port,
 						       flush_cb);
-			return -1;
+			disable_camera(dev);
+			return -EINVAL;
```

Comment:
Is EINVAL the most appropriate error here? Is the return value from vchiq_mmal_port_enable not suitable?

Reply:
> Is the return value from vchiq_mmal_port_enable not suitable?

Often not, although that's a different issue.

https://github.com/raspberrypi/linux/blob/rpi-5.10.y/drivers/staging/vc04_services/vchiq-mmal/mmal-vchiq.c#L1273 returns the MMAL error code without mapping it to Linux error codes, so you'll often see -3 [-MMAL_EINVAL](https://github.com/raspberrypi/userland/blob/master/interface/mmal/mmal_types.h#L51) appearing as [-ESRCH](https://elixir.bootlin.com/linux/latest/source/include/uapi/asm-generic/errno-base.h#L7)

Reply:
OK, fixed up the error handling and returning the error from vchiq_mmal_port_enable both here and at line 599.